File: union-find-forest/uff.py
class UFF:

    def __init__(self, n):
        self.a = list(range(n))
        self.r = [1] * n

    def find(self, x):
        z = x
        while True:
            y = self.a[z]
            if y == z:
                result = y
                break
            z = y
        while True:
            y = self.a[x]
            self.a[x] = result
            if y == result:
                break
            x = y
        return result

    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return x

        if self.r[x] > self.r[y]:
            self.a[y] = x
            return x
        elif self.r[x] < self.r[y]:
            self.a[x] = y
        else:
            self.a[x] = y
            self.r[x] += 1
        return y


if __name__ == '__main__':
    uff = UFF(8)
    uff.union(0, 1)
    uff.union(2, 3)
    uff.union(4, 5)
    uff.union(6, 7)
    uff.union(0, 2)

    for i in range(8):
        print(i, uff.find(i))

File: union-find-forest/UFF.java
import java.util.Scanner;

public class UFF {

    private int[] array;
    private int[] rank;

    public int find(int x){
        if (x == array[x]) return (x);
        return (array[x] = find(array[x]));
    }

    public String union (int x, int y){
        int X,Y;
        if ((X = find(x)) == (Y=find(y))) {
            return "YES";
        }
        if (rank[X] < rank[Y]) {
            array[X] = Y;
        } else{
            array[Y] = X;
            if (rank[X] == rank[Y]) {
                ++rank[X];
            }
        }
        return "NO";
    }

    public static void main(String[] args) {
        UFF unionFindForest = new UFF();
        Scanner scanner = new Scanner(System.in);
        String[] input= scanner.nextLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int k = Integer.parseInt(input[1]);
        unionFindForest.array = new int[n];
        unionFindForest.rank = new int[n];
        for (int i = 0; i < n; i++){
            unionFindForest.rank[i] = 1;
            unionFindForest.array[i] = i;
        }
        for (int i = 0; i < k; i++){
            input = scanner.nextLine().split(" ");
            System.out.println(unionFindForest.union(Integer.parseInt(input[0]), Integer.parseInt(input[1])));
        }
    }
}
